src/message/ping-parser.h

```diff
@@ -30,84 +44,139 @@ class PingParser
         ERROR, // Checksum didn't check out
     };
 
-    // variables used for parsing
+    /**
+     * @brief Reset parser state
+     *
+     */
+    void reset()
+    {
+        state_ = State::WAIT_START;
+    }
+
+    /**
+     * @brief Parse a single byte
+     *
+     * @param data
+     * @return State
+     */
+    PingParser::State parseByte(const uint8_t data);
+
+// variables used for parsing
 private:
     uint8_t* rxBuffer_;
     uint32_t rxBufferLength_;
     uint32_t rxCount_ = 0;
     uint16_t payloadLength_ = 0;
-    uint8_t state_ = WAIT_START;
+    State state_ = PingParser::State::WAIT_START;
+};
 
-public:
-    void reset()
-    {
-        state_ = WAIT_START;
-    }
+static const char* PINGPARSER_STATE_TO_STRING(PingParser::State state) {
+    switch(state) {
+        case PingParser::State::NEW_MESSAGE:
+            return "NEW_MESSAGE";
+        case PingParser::State::WAIT_START:
+            return "WAIT_START";
+        case PingParser::State::WAIT_HEADER:
+            return "WAIT_HEADER";
+        case PingParser::State::WAIT_LENGTH_L:
+            return "WAIT_LENGTH_L";
+        case PingParser::State::WAIT_LENGTH_H:
+            return "WAIT_LENGTH_H";
+        case PingParser::State::WAIT_MSG_ID_L:
+            return "WAIT_MSG_ID_L";
+        case PingParser::State::WAIT_MSG_ID_H:
+            return "WAIT_MSG_ID_H";
+        case PingParser::State::WAIT_SRC_ID:
+            return "WAIT_SRC_ID";
+        case PingParser::State::WAIT_DST_ID:
+            return "WAIT_DST_ID";
+        case PingParser::State::WAIT_PAYLOAD:
+            return "WAIT_PAYLOAD";
+        case PingParser::State::WAIT_CHECKSUM_L:
+            return "WAIT_CHECKSUM_L";
+        case PingParser::State::WAIT_CHECKSUM_H:
+            return "WAIT_CHECKSUM_H";
+        case PingParser::State::ERROR:
+            return "ERROR";
+        default:
+            return "UNKNOWN";
+    };
+}
```

Comment:
https://github.com/Neargye/magic_enum

Reply:
This should be embedded level compatible